YNU-354: Protocol Docs (#69)
layer-3/docs: docs/protocol/architecture.mdx

Reviewer: Please, review how mermaid diagrams are rendered in the dark mode, because chosen colours are not suited well

Author: I am trying out different color schemes that would go well with light and dark mode both

Author: I have kept the light mode as default and added a white background to it everywhere. So regardless of mode, it is legible

---
sidebar_position: 3
title: "Architecture"
---

import Tooltip from '@site/src/components/Tooltip';

# Architecture

## System Overview

The <Tooltip content="The on-chain smart contract protocol.">Nitrolite</Tooltip> protocol architecture consists of multiple layers working together to enable scalable, secure <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">state channel</Tooltip> operations:

```mermaid
graph TB
    subgraph Application["Application Layer"]
        APP[Chess, DEX, Gaming, Payments, Custom Logic]
    end
    
    subgraph OffChain["Off-Chain Layer (Fast Updates)"]
        CLIENT[Client SDK]
        BROKER[Clearnode]

>         RPC[Nitro RPC WebSocket]

Reviewer: We don't need to mention here websocket. I'm not sure whether we need to declare this entity here at all, as it's not referenced anywhere

Reviewer: Maybe that's just a typo, make sure to check how it is rendered on the web page

Author: ```mermaid
graph TB

  %% Layered architecture with legible fills for light/dark modes

  subgraph Application["Application Layer"]
    direction TB
    APP["Chess, DEX, Gaming, Payments, Custom Logic"]
  end

  subgraph OffChain["Off-Chain Layer (Fast Updates)"]
    direction LR
    CLIENT["Client SDK"]
    BROKER["Broker (Clearnode)"]
  end

  subgraph OnChain["On-Chain Layer (Smart Contracts)"]
    direction TB
    CONTRACTS["Custody / App Contracts"]
  end

  subgraph Blockchain["Blockchain Layer"]
    direction TB
    CHAIN["Ethereum, Polygon, etc."]
  end

  APP --> CLIENT
  APP --> BROKER
  CLIENT <-->|"Nitro RPC (WebSocket)"| BROKER
  CLIENT -. "Checkpoint / Settle" .-> CONTRACTS
  BROKER -. "Join / Verify" .-> CONTRACTS
  CONTRACTS --> CHAIN

  %% Subgraph fills (kept light for both themes)
  style Application fill:#e1f5ff,stroke:#9ad7ff,color:#111
  style OffChain fill:#fff4e1,stroke:#ffd497,color:#111
  style OnChain fill:#ffe1f5,stroke:#ffbde6,color:#111
  style Blockchain fill:#f0f0f0,stroke:#c9c9c9,color:#111

  %% Emphasize the RPC link (index 2 in the edge list above)
  linkStyle 2 stroke:#0ea5e9,stroke-width:2px
```

I have corrected it to this, NitroRPC is meant to be used to communicate between SDK and ClearNode.

Reviewer: Yeah, like this, but I'd still remove "(WebSocket)" part as it can depend on the implementation

Author: Done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

```

## Communication Patterns

### Channel Opening

The <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">channel</Tooltip> opening process follows a coordinated sequence between client and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.">a clearnode</Tooltip>:

1. Client requests channel creation from <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.">a clearnode</Tooltip> via <Tooltip content="The off-chain communication protocol.">Nitro RPC</Tooltip>
2. The clearnode returns signed initial <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data.">state</Tooltip>
3. Client submits transaction to <Tooltip content="The underlying distributed ledger technology (e.g., Ethereum).">blockchain</Tooltip>

> 4. The clearnode joins the channel on-chain

Reviewer: That's outdated, after Client submits initial state channel becomes open, and then clearnode just handles the `Opened` event and adds this channel as `open` on its side

Author: <img width="856" height="301" alt="image" src="https://github.com/user-attachments/assets/74309f0e-6e97-499b-979c-98ee5f3909bd" />
Is this correct? @philanton

Reviewer: Yep

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

5. Channel becomes active

```mermaid
sequenceDiagram
    participant Client
    participant Clearnode
    participant Blockchain
    
    Client->>Clearnode: create_channel request
    Clearnode->>Client: signed initial state
    Client->>Blockchain: submit create() transaction
    Blockchain->>Clearnode: Created event
    Clearnode->>Blockchain: submit join() transaction
    Blockchain->>Clearnode: Opened event
    Clearnode->>Client: channel_update notification
```

:::success Cooperative Opening
Channel opening requires cooperation between both parties, ensuring mutual agreement before funds are locked.
:::

### Off-Chain Updates

Off-chain updates occur entirely between <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">participants</Tooltip> without touching the blockchain:

> 1. Participants exchange signed <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data.">state</Tooltip> updates via <Tooltip content="The off-chain communication protocol.">Nitro RPC</Tooltip>

Reviewer: Actually, if we are talking about App Sessions, the exchange of states should be handled by App currently, and then, once App Session state has enough signatures to satisfy quorum, someone being responsible will need to submit this signed state to the Clearnode

Author: **Off-Chain Updates**:
1. Participants exchange signed state updates via Nitro RPC
   - *Note for App Sessions*: The exchange of states is handled by the App itself. Once the App Session state has enough signatures to satisfy quorum, a responsible party submits this signed state to the Clearnode.
   
   
Added it as a note, @philanton

Reviewer: Ahahha, welllll, it just sounds like))):
1. We handle this for you
  * Actually no, you should handle it by yourself

Author: Understood

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

2. No blockchain transactions required
3. Latest valid state maintained off-chain

> 4. Can be checkpointed on-chain at any time

Reviewer: In the ideal world - yes, but currently we don't store this off-chain state, because it's not related to the on-chain state, so we can't even checkpoint as well. But we are working on it, and I believe it will be more enforced in the next version of the protocol

Author: 4. Can be checkpointed on-chain at any time
   - *Current Implementation Note*: While this is the ideal design goal, the current implementation does not store off-chain state that is not related to on-chain state, so checkpointing is not currently supported. This functionality is under development and will be more enforced in the next version of the protocol.

Added this as a note here as well, @philanton

Reviewer: While this is the ideal design goal, ~~the current implementation does not store off-chain state that is not related to on-chain state~~ the current implementation does not store the state off-chain, so checkpointing is not currently supported. This functionality is under development and will be more enforced in the next version of the protocol.

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

:::tip Zero Gas Fees
Off-chain updates are instant (< 1 second) and incur zero gas fees, enabling high-frequency operations.
:::

### Channel Closing

Channels can be closed in two ways:

**Cooperative Closure**:
1. All participants sign final state
2. Any participant closes on-chain with signed final state
3. Fast and gas-efficient (1 transaction)

**Non-Cooperative Closure**:
1. Challenge-response mechanism resolves disputes
2. Requires challenge period for security
3. More expensive but works when parties disagree

```mermaid
graph LR
    A[Active Channel] --> B{Parties Agree?}
    B -->|Yes| C[Cooperative Close<br/>Fast, Cheap]
    B -->|No| D[Challenge-Response<br/>Slow, Secure]
    
    style C fill:#90EE90
    style D fill:#FFB6C1
```

## Fund Flow

The following diagram illustrates how funds flow through the Nitrolite protocol:

```mermaid
graph TB
    A[User Wallet<br/>ERC-20] -->|deposit| B[Custody Contract<br/>Available]

>     B -->|create/join| C[Channel<br/>Locked]

Reviewer: don't mention `join` here

Reviewer: But maybe `resize` should be here as well

Author: done

    C -->|off-chain updates| D[Unified Balance<br/>Clearnode]

>     D -->|app sessions| E[Virtual Channels<br/>Application]

Reviewer: I'd use `open/deposit app session` instead of `app sessions`

Author: done

---
sidebar_position: 3
title: "Architecture"
---

import Tooltip from '@site/src/components/Tooltip';

# Architecture

## System Overview

The <Tooltip content="The on-chain smart contract protocol.">Nitrolite</Tooltip> protocol architecture consists of multiple layers working together to enable scalable, secure <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">state channel</Tooltip> operations:

```mermaid

> graph TB

Reviewer 2: I would make the following changes:
1. Put "Custody, Adjudicator contracts" instead of "Custody / App contracts", as there is no App contracts on-chain.
2. Rename "Broker (Clearnode)" to just "Clearnode" as it is not only a broker (and to be honest, it is not a broker right now either), but a server running software to enable chain-abstraction.
3. Rename "Clearnode -> On-Chain Layer" arrow from "Join / Verify" to "Observe events" as currently Clearnode does NOT call the `Custody.join(...)` method - this process was simplified to a one-step `Custody.create(...)` from the user side.
4. Extract "Client SDK" from "Off-chain Layer" to be a separate component
5. Remove "Application layer -> Clearnode" arrow as all communication should go though Client SDK
6. Rename "Client SDK -> On-Chain layer" arrow from "Checkpoint / Settle" to "Operate on-chain state channels"
7. Rename "Client SDK -> Clearnode" arrow from "Nitro RPC (WebSocket)" to "Communicate via WebSocket using NitroRPC protocol"

I hope it all makes sense.

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

Reviewer 2: Great!
However, on my machine the "On-Chain Layer (Smart Contracts)" heading is cut as "On-Chain Layer (Smart", and the "Off-Chain Layer (Fast Updates)" is cut as "Off-Chain Layer (Fast".

  %% Layered architecture with legible fills for light/dark modes

  subgraph Application["Application Layer"]
    direction TB
    APP["Chess, DEX, Gaming, Payments, Custom Logic"]
  end

  subgraph OffChain["Off-Chain Layer (Fast Updates)"]
    direction LR
    CLIENT["Client SDK"]
    BROKER["Broker (Clearnode)"]
  end

  subgraph OnChain["On-Chain Layer (Smart Contracts)"]
    direction TB
    CONTRACTS["Custody / App Contracts"]
  end

  subgraph Blockchain["Blockchain Layer"]
    direction TB
    CHAIN["Ethereum, Polygon, etc."]
  end

  APP --> CLIENT
  APP --> BROKER
  CLIENT <-->|"Nitro RPC (WebSocket)"| BROKER
  CLIENT -. "Checkpoint / Settle" .-> CONTRACTS
  BROKER -. "Join / Verify" .-> CONTRACTS
  CONTRACTS --> CHAIN

  %% Subgraph fills (kept light for both themes)
  style Application fill:#e1f5ff,stroke:#9ad7ff,color:#111
  style OffChain fill:#fff4e1,stroke:#ffd497,color:#111
  style OnChain fill:#ffe1f5,stroke:#ffbde6,color:#111
  style Blockchain fill:#f0f0f0,stroke:#c9c9c9,color:#111

  %% Emphasize the RPC link (index 2 in the edge list above)
  linkStyle 2 stroke:#0ea5e9,stroke-width:2px
```

## Communication Patterns

> ### Channel Opening

Reviewer 2: ```suggestion
### On-Chain Channel Opening
```

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

The <Tooltip content="A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.">channel</Tooltip> opening process follows a coordinated sequence between client and <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.">a clearnode</Tooltip>:

1. Client requests channel creation from <Tooltip content="A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.">a clearnode</Tooltip> via <Tooltip content="The off-chain communication protocol.">Nitro RPC</Tooltip>

> 2. The clearnode returns signed initial <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data.">state</Tooltip> (signature at index 1)

Reviewer 2: Please change "The clearnode returns signed initial ..." to "The clearnode returns a channel struct and a signed initial ..." with a tooltip for the "channel".

Reviewer 2: And please update the diagram below accordingly.

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

3. Client signs the initial <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data.">state</Tooltip> (signature at index 0)

> 4. Client submits transaction to <Tooltip content="The underlying distributed ledger technology (e.g., Ethereum).">blockchain</Tooltip> with both signatures

Reviewer 2: ```suggestion
4. Client calls the `create(...)` method of the Custody Smart Contarct on the <Tooltip content="The underlying distributed ledger technology (e.g., Ethereum).">blockchain</Tooltip>, providing the channel and initial state with both signatures
```

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

5. Contract verifies signatures and emits `Opened` event
6. Channel becomes ACTIVE immediately
7. The clearnode monitors the `Opened` event and updates its internal state

```mermaid
sequenceDiagram
    participant Client
    participant Clearnode
    participant Blockchain
    
    Client->>Clearnode: create_channel request
    Clearnode->>Client: signed initial state (clearnode signature)
    Client->>Client: Add own signature
    Client->>Blockchain: create() with BOTH signatures
    Blockchain->>Blockchain: Verify signatures<br/>Set status = ACTIVE
    Blockchain->>Blockchain: Emit Opened event
    Blockchain-->>Clearnode: Opened event (monitored)
    Clearnode->>Client: channel_update notification
```

:::success Cooperative Opening
Channel opening requires cooperation between both parties, ensuring mutual agreement before funds are locked.
:::

### Off-Chain Updates

**Off-Chain Updates**:

> 1. <Tooltip content="An entity (identified by a wallet address) that is part of a channel.">Participants</Tooltip> exchange signed <Tooltip content="A snapshot of the channel at a point in time, including fund allocations and application-specific data.">state</Tooltip> updates via <Tooltip content="The off-chain communication protocol.">Nitro RPC</Tooltip>

Reviewer 2: I have noticed that a "state" with the same tooltip is used interchangeably across on-chain and off-chain channels. I guess this creates confusion for readers making them think this is the same structure. I think this can be fixed by differentiating those terms as "state channel state" / "on-chain state" and "app session state" / "off-chain state".
@ihsraham @dimast-x @philanton , WDYT?

Reviewer: I agree, they should be different, but let's not allocate too much on it right now), off-chain vs on-chain state is good to me

2. No blockchain transactions required

3. Latest valid state maintained off-chain

4. Can be checkpointed on-chain at any time

   - *Current Implementation Note*: While this is the ideal design goal, the current implementation does not store off-chain state that is not related to on-chain state, so checkpointing is not currently supported. This functionality is under development and will be more enforced in the next version of the protocol.

:::tip Zero Gas Fees
Off-chain updates are instant (< 1 second) and incur zero gas fees, enabling high-frequency operations.
:::

> ### Channel Closing

Reviewer 2: ```suggestion
### On-Chain Channel Closing
```

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

Channels can be closed in two ways:

> **Cooperative Closure**:

Reviewer 2: I think it is better to describe actual steps of the process if numerated lists are used.
Thus, could you please change the lists so they represent the step-by-step process more, while extracting comments on speed and gas efficiency as separate sentences, and not the part of the list?

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

1. All participants sign final state
2. Any participant closes on-chain with signed final state
3. Fast and gas-efficient (1 transaction)

**Non-Cooperative Closure**:
1. Challenge-response mechanism resolves disputes
2. Requires challenge period for security
3. More expensive but works when parties disagree

```mermaid
graph LR
    A[Active Channel] --> B{Parties Agree?}
    B -->|Yes| C[Cooperative Close<br/>Fast, Cheap]
    B -->|No| D[Challenge-Response<br/>Slow, Secure]
    
    style C fill:#90EE90,stroke:#333,color:#111
    style D fill:#FFB6C1,stroke:#333,color:#111
```

## Fund Flow

The following diagram illustrates how funds flow through the Nitrolite protocol:

```mermaid

> graph TB

Reviewer 2: Can you make the following changes, please:
1. Move the "Unified Balance Clearnode -> Custody Contract Available" arrow to "Channel Locked -> Custody Contract Available" and rename to "resize / challenge / close channel".
2. Rename the "App Sessions Application -> Unified Balance Clearnode" arrow from "close session" to "withdraw / close session".
3. Rename all boxes from "Text \n text" to "Funds type name \n (component name)" i.e. from "Unified Balance \n Clearnode" and "Channel \n Locked" to "Unified Balance \n (Clearnode)" and "Channel-Locked (Custody SC)".

And please update the "Flow Explanation" below accordingly.

Author: [60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f](https://github.com/layer-3/docs/pull/69/commits/60c17be9f05f54649ab1c1d8cd0c8b2664ac7c0f)

:::

### On-Chain Channel Closing

Channels can be closed in two ways:

**Cooperative Closure**:
1. All <Tooltip content={tooltipDefinitions.participant}>participants</Tooltip> negotiate and agree on the final <Tooltip content={tooltipDefinitions.channelState}>state</Tooltip>
2. Each participant signs the final state with `intent = FINALIZE`
3. Any participant submits the fully-signed final state to the <Tooltip content={tooltipDefinitions.custodyContract}>Custody Contract</Tooltip> via `close()`
4. Contract verifies all signatures and distributes funds according to final allocations
5. Channel status becomes FINAL

This is the preferred closure method. It requires only 1 transaction and is gas-efficient.

**Non-Cooperative Closure**:
1. A participant submits the latest known state to the <Tooltip content={tooltipDefinitions.custodyContract}>Custody Contract</Tooltip> via `challenge()`
2. Contract verifies signatures and sets channel status to DISPUTE
3. A challenge period begins (e.g., 24 hours), allowing the other party to respond
4. If a newer valid state is submitted during the challenge period, it replaces the current state

> 5. After the challenge period expires, any participant calls `close()` to finalize with the latest submitted state

Reviewer 2: Let's clarify what happens if a newer state is submitted during challenge period

```suggestion
4. If participants decides to cooperate again, they may produce a newer valid state, and any of them can submit it via `checkpoint()`, thus stopping the challenge period and moving the channel from DISPUTE back to ACTIVE status
5. If not, after the challenge period expires, any participant calls `close()` to finalize with the latest submitted state
```

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)

```mermaid
graph TB
    A[User Wallet<br/>ERC-20] -->|deposit| B[Available<br/>Custody Contract]
    B -->|create/resize| C[Channel-Locked<br/>Custody SC]
    C -->|off-chain updates| D[Unified Balance<br/>Clearnode]
    D -->|open/deposit| E[App Sessions<br/>Application]
    E -->|withdraw / close session| D

>     C -->|resize / challenge / close channel| B

Reviewer 2: Below are some suggestions.
Also, if I understand correctly, in v0.5.0 if users behave strictly by the protocol, their funds should never reach the "Channel-locked" state. Thus, I suggest we specify that funds are moved directly from Custody-locked to Unified balance, but still add possibility to reach the "Channel-locked" state by the resize:

```suggestion
    A[User Wallet<br/>#40;ERC-20#41;] -->|deposit| B[Available<br/>#40;Custody SC#41;]
    B -->|resize| D[Unified Balance<br/>#40;Clearnode#41;]
    B <-->|resize| C[Channel-Locked<br/>#40;Custody SC#41;]
    C <-->|resize| D
    D -->|open/deposit| E[App Sessions<br/>#40;Application#41;]
    E -->|withdraw / close session| D
    D -->|resize / close channel| B
    B -->|withdraw| A
```

And, please, update the textual description of steps below accordingly.
P.S. "#40;" and "#41;" stand for "(" and ")" respectively.

<img width="1067" height="1421" alt="Image" src="https://github.com/user-attachments/assets/845b8811-4f3a-4542-a632-ec3720de0333" />

Author: [aad0b80ca8e5c74d4c329a9ed21b7db470b0d336](https://github.com/layer-3/docs/pull/69/commits/aad0b80ca8e5c74d4c329a9ed21b7db470b0d336)
